### PR TITLE
Add pidfdinfo and SocketFDInfo

### PR DESCRIPTION
**Pull request by dalance**
I added `pidfdinfo` and `SocketFDInfo`.

### CODE REVIEW

File: src/libproc/proc_pid.rs [hunk 1]
```diff
@@ -637,3 +637,452 @@ impl From<uint32_t> for ProcFDType {
...
+///     // Open TCP port:8000 to test.
+///     let _listener = TcpListener::bind("127.0.0.1:8000");
+///
+///     if let Ok(info) = pidinfo::<BSDInfo>(pid, 0) {
+///         if let Ok(fds) = listpidinfo::<ListFDs>(pid, info.pbi_nfiles as usize) {
+///             for fd in &fds {
+///                 match fd.proc_fdtype.into() {
+///                     ProcFDType::Socket => {
+///                         if let Ok(socket) = pidfdinfo::<SocketFDInfo>(pid, fd.proc_fd) {
+///                             match socket.psi.soi_kind.into() {
```
Comment (andrewdavidmackenzie) [inline src/libproc/proc_pid.rs]: You could reduce indentation one layer here :-) by matching Ok(SocketInfo...) if you wanted....

File: src/libproc/proc_pid.rs [hunk 2]
```diff
@@ -637,3 +637,452 @@ impl From<uint32_t> for ProcFDType {
...
+///                                     let mut port = 0;
+///                                     port |= info.tcpsi_ini.insi_lport >> 8 & 0x00ff;
+///                                     port |= info.tcpsi_ini.insi_lport << 8 & 0xff00;
+///
+///                                     // access to the member of `insi_laddr` is unsafe becasuse of union type.
+///                                     let s_addr = unsafe { info.tcpsi_ini.insi_laddr.ina_46.i46a_addr4.s_addr };
+///
+///                                     // change endian because insi_laddr is network endian.
+///                                     let mut addr = 0;
+///                                     addr |= s_addr >> 24 & 0x000000ff;
```
Comment (andrewdavidmackenzie) [inline src/libproc/proc_pid.rs]: I've never used it, but maybe rust has something to offer you to make endianness related stuff more idiomatic?
https://docs.rs/byteorder/1.3.1/byteorder/ maybe?

File: src/libproc/proc_pid.rs [hunk 3]
```diff
@@ -637,3 +637,452 @@ impl From<uint32_t> for ProcFDType {
...
+#[test]
+fn pidfdinfo_test() {
+    use std::process;
+    use std::net::TcpListener;
+    let pid = process::id() as i32;
+
+    let _listener = TcpListener::bind("127.0.0.1:65535");
+
+    match pidinfo::<BSDInfo>(pid, 0) {
+        Ok(info) => {
```
Comment (andrewdavidmackenzie) [inline src/libproc/proc_pid.rs]: could simplify by using unwrap(), or expect() in the tests - no problem though, just mentioning it.

File: src/libproc/proc_pid.rs [hunk 4]
```diff
@@ -637,3 +637,452 @@ impl From<uint32_t> for ProcFDType {
...
+                },
+                Err(err) => assert!(false, "Error retrieving process info: {}", err)
+            }
+        },
+        Err(err) => assert!(false, "Error retrieving process info: {}", err)
+    };
+}
+
+#[repr(C)]
+#[derive(Default)]
```
Comment (andrewdavidmackenzie) [inline src/libproc/proc_pid.rs]: Do you actually need Default on all of these structs?

File: src/libproc/proc_pid.rs [hunk 5]
```diff
@@ -637,3 +637,452 @@ impl From<uint32_t> for ProcFDType {
...
+    Ndrv      = 4,
+    /// Kernel Event Sockets
+    KernEvent = 5,
+    /// Kernel Control Sockets
+    KernCtl   = 6,
+    /// Unknown
+    Unknown,
+}
+
+impl From<c_int> for SocketInfoKind {
```
Comment (andrewdavidmackenzie) [inline src/libproc/proc_pid.rs]: It's a pity rust can't derive this type of stuff for us, to go from value -> enum, like you can in Java...

File: src/libproc/proc_pid.rs [hunk 6]
```diff
@@ -637,3 +637,452 @@ impl From<uint32_t> for ProcFDType {
...
+    pub sbi_hiwat: uint32_t,
+    pub sbi_mbcnt: uint32_t,
+    pub sbi_mbmax: uint32_t,
+    pub sbi_lowat: uint32_t,
+    pub sbi_flags: c_short,
+    pub sbi_timeo: c_short,
+}
+
+#[repr(C)]
+pub union SocketInfoProto {
```
Comment (andrewdavidmackenzie) [inline src/libproc/proc_pid.rs]: never used a union in rust!

**Comment by andrewdavidmackenzie** [comment]
File size has grown a lot with all the new stuff related to FD info.... could consider adding in a separate file even?

All the code looks good. Let me know if you make any changes as a result of comments, or not. I'm happy to merge as is.

File: src/libproc/proc_pid.rs (hunk 4 above)
Comment (dalance) [reply src/libproc/proc_pid.rs]: This Default is used in `pidfdinfo` like `let mut pidinfo = T::default();`.
I think all members of `SocketFDInfo` must have Default.

File: src/libproc/proc_pid.rs (hunk 5 above)
Comment (dalance) [reply src/libproc/proc_pid.rs]: Yes, I think so...

File: src/libproc/proc_pid.rs (hunk 3 above)
Comment (dalance) [reply src/libproc/proc_pid.rs]: I simplified it.

**Comment by dalance** [comment]
I think the file could be separated to files having a C function like below.

proc_pid.rs -> only export for proc_pid namespace
proc_pidinfo.rs
proc_pidfdinfo.rs
...

Anyway, file structure change is big issue.
I think another PR is rquired for it.
Could you merge this?

**Comment by andrewdavidmackenzie** [comment]
OK, merged.
Do you want to take a shot at splitting up into multiple files?
If not, we can create an issue and I can try to get to it sometime.

**Comment by dalance** [comment]
Thanks!
I don't want it.

I have another PR to add proc_pid_rusage.
Should I wait until file separation complete?

**Pull request merged**